lychee-lib/src/types/basic_auth.rs

```diff
@@ -0,0 +1,174 @@
+    TooManyParts(usize),
+
+    #[error("Basic auth credentials error")]
+    BasicAuthCredentialsParseError(#[from] BasicAuthCredentialsParseError),
+
+    #[error("Regex compile error")]
+    RegexError(#[from] regex::Error),
+}
+
+/// [`BasicAuthSelector`] provides basic auth credentials for URLs which match
+/// the specified regex. This allows users to set different credentials based
+/// on the URLs they want to target.
+#[derive(Debug, Clone, DeserializeFromStr, PartialEq)]
+pub struct RawBasicAuthSelector {
+    /// The basic auth credentials made up of username and password
+    pub credentials: BasicAuthCredentials,
+
+    /// This regex matches URLs which will use the basic auth credentials
+    pub url: String,
+}
+
+impl FromStr for RawBasicAuthSelector {
+    type Err = BasicAuthSelectorParseError;
+
+    fn from_str(selector: &str) -> Result<Self, Self::Err> {
+        let parts: Vec<_> = selector.trim().split(' ').collect();
+
+        if parts.len() == 1 {
+            return Err(BasicAuthSelectorParseError::MissingCredentials);
+        }
+
+        if parts.len() != 2 {
+            return Err(BasicAuthSelectorParseError::TooManyParts(parts.len()));
+        }
+
+        Ok(Self {
+            credentials: parts[1].parse()?,
+            url: parts[0].to_string(),
+        })
+    }
+}
+
+/// [`BasicAuthSelector`] provides basic auth credentials for URLs which match
+/// the specified regex. This allows users to set different credentials based
+/// on the URLs they want to target. The basic auth username and password will
+/// be used by the URL matched by the compiled regex.
+#[derive(Clone, Debug)]
+pub struct BasicAuthSelector {
+    /// The basic auth credentials made up of username and password
+    pub credentials: Authorization<Basic>,
+
+    /// This regex matches URLs which will use the basic auth credentials
+    pub url: Regex,
+}
+
+impl FromStr for BasicAuthSelector {
+    type Err = BasicAuthSelectorParseError;
+
+    fn from_str(input: &str) -> Result<Self, Self::Err> {
+        let selector: RawBasicAuthSelector = input.parse()?;
+        BasicAuthSelector::try_from(selector)
+    }
+}
+
+impl TryFrom<RawBasicAuthSelector> for BasicAuthSelector {
+    type Error = BasicAuthSelectorParseError;
+
+    fn try_from(selector: RawBasicAuthSelector) -> Result<Self, Self::Error> {
+        Self::try_from(&selector)
+    }
+}
+
+impl TryFrom<&RawBasicAuthSelector> for BasicAuthSelector {
+    type Error = BasicAuthSelectorParseError;
+
+    fn try_from(selector: &RawBasicAuthSelector) -> Result<Self, Self::Error> {
+        Ok(Self {
+            credentials: Authorization::basic(
+                &selector.credentials.username,
+                &selector.credentials.password,
+            ),
+            url: Regex::new(&selector.url)?,
+        })
+    }
+}
+
+#[cfg(test)]
+mod test {
+    use super::*;
+
+    #[test]
```

Comment:
We should add some negative tests as well.

Reply:
Yes, you are right. I will add those.

Reply:
I added multiple negative tests. This also resolves the above comment.